### ISSUE TITLE
Functional components fail to transition without a unique tag and key

### ISSUE THREAD

**Issue by robmadole** [vuejs/vue#11110]
### Version
2.6.11

### Reproduction link
[https://github.com/robmadole/vue-func-comp-transition](https://github.com/robmadole/vue-func-comp-transition)

### Steps to reproduce

- npm install
- npm run dev
- Open your browser and click the toggle button

### What is expected?
At first, I expected the transition to work.

Now, I think a warning in the dev console would be best and some docs that transitioning functional components is tricky and a link how to do it (similar to how other transition-related warnings are presented)

### What is actually happening?
Because the functional component do not have a unique tag and key the <transition> component doesn't operate on them. It can't distinguish between them so it assumes no transition is needed. Essentially it doesn't think the children have changed.

---
 I think I now understand why this occurs after a bit of source code reading. Would it be possible to detect that two functional components are being transitioned and that Vue cannot distinguish between the two? At the very least, a section in the docs to warn against this?

We ran into this quite organically; trying to transition two functional components. It didn't work as expected (we've used transitions before) and I spun off for about an hour to figure out why.

**Comment by posva** [comment]
I'm not sure if this is a bug: functional component not having an instance, they look more like raw `div` elements, in which case, not having a key will not trigger a transition because Vue reuses the element (this is in the doc), which is what you are observing.

The reason a `key` doesn't seem to work is because functional components do not inherit attributes by default, it has to be done manually:

```js
export default {
      functional: true,
      render: (h, context) => h("div", { ...context.data }, "One")
}
```

When using templates, it also has to be explicitly passed:

```html
<div :key="data.key">One</div>
```

**Comment by robmadole** [comment]
@posva no, it is certainly not a bug. I figured out _why_ it just took me a long time. It's just something that tripped me up so an improvement would be to either document it a little more (I missed it) or perhaps mark it [similar to how transition-group warns of missing keys](https://github.com/vuejs/vue/blob/8a2dbf50105ea729125a42fecfe2c2f0371d7836/src/platforms/web/runtime/components/transition-group.js#L72)? Is that possible?

**Comment by posva** [comment]
> perhaps mark it similar to how transition-group warns of missing keys? Is that possible?

That wouldn't be possible because the warning in transition-group always appear. Here we need some way of deciding when to show the warning and even in those cases, reusing nodes could be intentional.

Let's move this to docs then
